File: 2_implementacion_algoritmos/algorithms/quicksort.cpp
#include "quicksort.h"
#include <bits/stdc++.h>
#include <utility>
using namespace std;

Quicksort::Quicksort(){}

/* Toma el ultimo elemento como pivote,
   localiza el pivote en su correcta posicion en el array ordenado
   localiza los elemento mas pequeños a la izquierda del pivote
   localiza los elementos mas grande a la derecha del pivote
   */
int Quicksort::partition(int arr[], int low, int high)
{
	//pivote
	int pivot  = arr[high]; 

	//indice del elemento mas pequenio e indica la correcta posicion del pivote encontrado
	int i  = (low - 1);

	for (int j = low; j <= high -1; j++)
	{
		//si el elemento es mas pequenio que el pivote
		if (arr[j] < pivot)
		{
			i++; //incrementamos el indice del elemento mas pequenio
			swap(arr[i], arr[j]); 
		}
	}
	swap(arr[i+1], arr[high]);
	return (i+1);
}

void Quicksort::quicksort(int arr[], int low, int high)
{
	if (low < high)
	{
		int pi = partition(arr, low, high);

		//ordenar los elementos antes y despues del pivote
		quicksort(arr, low, pi - 1);
		quicksort(arr, pi+1, high);
	}
}

void Quicksort::printArray(int arr[], int size)
{

	int i;
	for (i=0; i<size; i++)
		cout << arr[i] << " "; 
	cout << endl;
} 

/*
int main(){
	int arrayint[500];
	memset(arrayint,0,500);
	string line;
	std::ifstream ifs;
	ifs.open("generatedTestData/example_500.txt", std::ifstream::in);
	int idx = 0;
	if(ifs.is_open()){
		while(getline(ifs, line))
		{
			//cout << line << "\n";
			arrayint[idx] = stoi(line);
			idx++;
		}
		ifs.close();
		quicksort(arrayint, 0, 500-1);
		printArray(arrayint,500);

	}
	else cout << "No se pudo abrir el archivo";

	return 0;
}*/

File: 2_implementacion_algoritmos/algorithms/quicksort.py
class QuickSort:
    def __init__(self):
        print("constructor")
    
    #Puntos de inicio y final y el ultimo elemento del array
    def partition(self,start, end, array):
        #Inicializando indice del pivote inicial
        pivot_index = start
        pivot = array[pivot_index]

        #iterar hasta que el ptro de inicio cruce con el ptro final,
        #intercambiamos el pivote con el elemento sobre el ptro final
        while start < end:

            #incrementamos el ptro de inicio hasta que encuentre
            #un elemento mas grande que el pivote
            while start < len(array) and array[start] <= pivot:
                start += 1

            #decrementamos el puntero final hasta que que encuentre
            #un elemento menos que el pivote
            while array[end] > pivot:
                end -= 1

            #Si el start y end no se han cruzado, intercambiamos los números
            #del ptro start y end
            if (start < end):
                array[start], array[end] = array[end], array[start]

        #Cambiamos el pivote con el elemento sobre el ptro final
        #Esto colocal el pivote en el correcto lugar de orden
        array[end], array[pivot_index] = array[pivot_index], array[end]

        #retorna el ptro final para dividir el array en 2
        return end

    def quicksort(self,start, end, array):
        if (start < end):
            #p es el indice de particion, array[p] es el correcto lugar
            p = self.partition(start, end, array)

            #ordenar los elemento antes y despues del pivote
            self.quicksort(start, p-1, array)
            self.quicksort(p+1, end, array)

